_posts/2020-02-06-xcode-selection.md

```diff
@@ -0,0 +1,113 @@
+ 
+```python
+xcode_version(
+    name = ‘xcode10.2’,
+    version = ‘10.2.1’
+)
+
+xcode_version(
+    name = ‘xcode11.1’,
+    version = ‘11.1’
+)
+
+available_xcodes(
+    name = ‘local_xcodes’,
+    default = ‘xcode10.2’,
+    versions = [‘:xcode10.2’ ,’:xcode11.1’]
+)
+
+available_xcodes(
+    name = ‘remote_xcodes’,
+    default = ‘xcode11.1’,
+    versions = [,’:xcode11.1’]
+)
+```
+
+Does that look familiar?
+
+
+We then modified the `xcode_config` rule to accept two `available_xcodes` dependencies to represent the locally and remotely available Xcode versions. Note that this new shape doesn’t accept the original attributes, `default` and `versions`. These concepts are reflected in the `available_xcodes` dependencies.
+
+```python
+xcode_config(
+    name = ‘my_config’,
+    local_versions = ‘:local_xcodes’,
+    remote_versions = ‘remote_xcodes’
+)
+```
+
+## Xcode version selection
+Both `xcode_config` shapes select the value of `--xcode_version` if it’s present, or else the default. They differ, however, in what constitutes the default, and in the behavior if the selected Xcode is only present on a single platform.
+
+You might remember that we were using the remote `xcode_config` with dynamic execution, which resulted in build failures if the default (remote) version wasn’t available locally. We addressed this issue firstly by having the new `xcode_config` set a mutually available Xcode version as the default if possible (otherwise the default of the `local_versions` target), and secondly by allowing selection of any Xcode version present in `local_versions` or `remote_versions`.
+
+A mutually available Xcode should result in the best performance, since it enables both local and remote execution, and since some actions (e.g. Swift compiles) must be executed locally, the local default is the next best thing. Xcode version selection occurs in this order, mutual first, then local default, but you can skip straight to the local default by passing `--experimental_prefer_mutual_xcode=false`.
+
+Since the new `xcode_config` requires both dependencies to be set, we will never default to a remote-only Xcode version. We considered allowing a remote-only Xcode version in the absence of a locally available Xcode, but decided that preventing builds from failing cryptically if there was no Xcode to execute local actions was more important than providing the flexibility for dynamic execution to behave like a purely remote strategy.
+
+## Configuring dynamic execution
+The last piece is having Bazel execute actions in the right location based on the availability of the selected Xcode. If we’ve selected a local- or remote-only Xcode, either via `--xcode_version` or by accepting the default, we want to keep the dynamic scheduler from trying to execute Xcode-related actions on the other system. 
```

Comment:
You might as well say "local-only or remote-only".   Factoring out the |only| is a minor savings for a probably readability hit.

Reply:
Done.